resource/resourceadapters/opener.go

```diff
@@ -38,13 +60,18 @@ func newInternalResourceOpener(st ResourceOpenerState, unitName string) (opener
 
 	var resourceClient ResourceRetryClientGetterFn
 
-	curl, _ := unit.CharmURL()
+	charmURL := (*charm.URL)(nil)
```

Comment:
Maybe not worth it, but could have an interface with CharmURL() and Tag() (which both app and unit implement) and pass to an extracted function of the code below to avoid the if unit != nil stuff

Reply:
ahaha I did this to begin with then deleted it because I thought you of all people wouldn't like it.